File: backend/app/api/risks.py
"""
Risk API - 风险/阻塞管理
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel
from ..database import get_db
from ..models.user import User
from ..models.stage_instance import TopicRisk, RiskLevel
from ..services.auth import get_current_user

router = APIRouter()


class RiskCreate(BaseModel):
    topic_id: int
    stage_instance_id: Optional[int] = None
    level: str = "yellow"  # red, yellow, green
    blocker_type: str
    blocker_name: Optional[str] = None
    description: str
    expected_resolve_date: Optional[datetime] = None
    owner_id: Optional[int] = None


class RiskUpdate(BaseModel):
    level: Optional[str] = None
    blocker_type: Optional[str] = None
    blocker_name: Optional[str] = None
    description: Optional[str] = None
    expected_resolve_date: Optional[datetime] = None
    owner_id: Optional[int] = None
    is_resolved: Optional[bool] = None
    resolution_note: Optional[str] = None


class RiskResponse(BaseModel):
    id: int
    topic_id: int
    stage_instance_id: Optional[int] = None
    level: str
    blocker_type: str
    blocker_name: Optional[str] = None
    description: str
    expected_resolve_date: Optional[datetime] = None
    owner_id: Optional[int] = None
    owner: Optional[dict] = None
    is_resolved: bool
    resolved_at: Optional[datetime] = None
    resolution_note: Optional[str] = None
    created_by_id: Optional[int] = None
    created_by: Optional[dict] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


def serialize_user(user):
    if not user:
        return None
    return {"id": user.id, "name": user.name, "email": user.email}


@router.get("/topics/{topic_id}/risks", response_model=List[RiskResponse])
def list_risks(
    topic_id: int,
    include_resolved: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取课题的所有风险"""
    query = db.query(TopicRisk).options(
        joinedload(TopicRisk.owner),
        joinedload(TopicRisk.created_by)
    ).filter(TopicRisk.topic_id == topic_id)
    
    if not include_resolved:
        query = query.filter(TopicRisk.is_resolved == False)
    
    risks = query.order_by(
        TopicRisk.is_resolved,
        TopicRisk.level.desc(),
        TopicRisk.created_at.desc()
    ).all()
    
    return [
        {
            **r.__dict__,
            "level": r.level.value if r.level else "yellow",
            "owner": serialize_user(r.owner),
            "created_by": serialize_user(r.created_by),
        }
        for r in risks
    ]


@router.post("/risks", response_model=RiskResponse)
def create_risk(
    data: RiskCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """创建风险"""
    risk = TopicRisk(
        topic_id=data.topic_id,
        stage_instance_id=data.stage_instance_id,
        level=RiskLevel(data.level) if data.level else RiskLevel.YELLOW,
        blocker_type=data.blocker_type,
        blocker_name=data.blocker_name,
        description=data.description,
        expected_resolve_date=data.expected_resolve_date,
        owner_id=data.owner_id,
        created_by_id=current_user.id
    )
    db.add(risk)
    db.commit()
    db.refresh(risk)
    
    risk = db.query(TopicRisk).options(
        joinedload(TopicRisk.owner),
        joinedload(TopicRisk.created_by)
    ).filter(TopicRisk.id == risk.id).first()
    
    return {
        **risk.__dict__,
        "level": risk.level.value if risk.level else "yellow",
        "owner": serialize_user(risk.owner),
        "created_by": serialize_user(risk.created_by),
    }


@router.put("/risks/{risk_id}", response_model=RiskResponse)
def update_risk(
    risk_id: int,
    data: RiskUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """更新风险"""
    risk = db.query(TopicRisk).filter(TopicRisk.id == risk_id).first()
    if not risk:
        raise HTTPException(status_code=404, detail="Risk not found")
    
    if data.level is not None:
        risk.level = RiskLevel(data.level)
    if data.blocker_type is not None:
        risk.blocker_type = data.blocker_type
    if data.blocker_name is not None:
        risk.blocker_name = data.blocker_name
    if data.description is not None:
        risk.description = data.description
    if data.expected_resolve_date is not None:
        risk.expected_resolve_date = data.expected_resolve_date
    if data.owner_id is not None:
        risk.owner_id = data.owner_id
    if data.is_resolved is not None:
        risk.is_resolved = data.is_resolved
        if data.is_resolved:
            risk.resolved_at = datetime.utcnow()
    if data.resolution_note is not None:
        risk.resolution_note = data.resolution_note
    
    db.commit()
    db.refresh(risk)
    
    risk = db.query(TopicRisk).options(
        joinedload(TopicRisk.owner),
        joinedload(TopicRisk.created_by)
    ).filter(TopicRisk.id == risk.id).first()
    
    return {
        **risk.__dict__,
        "level": risk.level.value if risk.level else "yellow",
        "owner": serialize_user(risk.owner),
        "created_by": serialize_user(risk.created_by),
    }


@router.delete("/risks/{risk_id}")
def delete_risk(
    risk_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """删除风险"""
    risk = db.query(TopicRisk).filter(TopicRisk.id == risk_id).first()
    if not risk:
        raise HTTPException(status_code=404, detail="Risk not found")
    
    db.delete(risk)
    db.commit()
    return {"message": "Risk deleted"}

File: src/api/risks.ts
import client from './client';

export interface Risk {
  id: number;
  topicId: number;
  stageInstanceId?: number;
  level: 'red' | 'yellow' | 'green';
  blockerType: string;
  blockerName?: string;
  description: string;
  expectedResolveDate?: string;
  ownerId?: number;
  owner?: { id: number; name: string; email: string };
  isResolved: boolean;
  resolvedAt?: string;
  resolutionNote?: string;
  createdById?: number;
  createdBy?: { id: number; name: string; email: string };
  createdAt: string;
  updatedAt: string;
}

export interface RiskCreate {
  topicId: number;
  stageInstanceId?: number;
  level?: string;
  blockerType: string;
  blockerName?: string;
  description: string;
  expectedResolveDate?: string;
  ownerId?: number;
}

export interface RiskUpdate {
  level?: string;
  blockerType?: string;
  blockerName?: string;
  description?: string;
  expectedResolveDate?: string;
  ownerId?: number;
  isResolved?: boolean;
  resolutionNote?: string;
}

function normalizeRisk(r: any): Risk {
  return {
    id: r.id,
    topicId: r.topic_id ?? r.topicId,
    stageInstanceId: r.stage_instance_id ?? r.stageInstanceId,
    level: r.level,
    blockerType: r.blocker_type ?? r.blockerType,
    blockerName: r.blocker_name ?? r.blockerName,
    description: r.description,
    expectedResolveDate: r.expected_resolve_date ?? r.expectedResolveDate,
    ownerId: r.owner_id ?? r.ownerId,
    owner: r.owner,
    isResolved: r.is_resolved ?? r.isResolved ?? false,
    resolvedAt: r.resolved_at ?? r.resolvedAt,
    resolutionNote: r.resolution_note ?? r.resolutionNote,
    createdById: r.created_by_id ?? r.createdById,
    createdBy: r.created_by ?? r.createdBy,
    createdAt: r.created_at ?? r.createdAt,
    updatedAt: r.updated_at ?? r.updatedAt,
  };
}

export const riskApi = {
  list: async (topicId: number, includeResolved = false): Promise<Risk[]> => {
    const response = await client.get<any[]>(`/topics/${topicId}/risks`, {
      params: { include_resolved: includeResolved }
    });
    return response.data.map(normalizeRisk);
  },

  create: async (data: RiskCreate): Promise<Risk> => {
    const response = await client.post<any>('/risks', {
      topic_id: data.topicId,
      stage_instance_id: data.stageInstanceId,
      level: data.level || 'yellow',
      blocker_type: data.blockerType,
      blocker_name: data.blockerName,
      description: data.description,
      expected_resolve_date: data.expectedResolveDate,
      owner_id: data.ownerId,
    });
    return normalizeRisk(response.data);
  },

  update: async (id: number, data: RiskUpdate): Promise<Risk> => {
    const payload: any = {};
    if (data.level !== undefined) payload.level = data.level;
    if (data.blockerType !== undefined) payload.blocker_type = data.blockerType;
    if (data.blockerName !== undefined) payload.blocker_name = data.blockerName;
    if (data.description !== undefined) payload.description = data.description;
    if (data.expectedResolveDate !== undefined) payload.expected_resolve_date = data.expectedResolveDate;
    if (data.ownerId !== undefined) payload.owner_id = data.ownerId;
    if (data.isResolved !== undefined) payload.is_resolved = data.isResolved;
    if (data.resolutionNote !== undefined) payload.resolution_note = data.resolutionNote;

    const response = await client.put<any>(`/risks/${id}`, payload);
    return normalizeRisk(response.data);
  },

  delete: async (id: number): Promise<void> => {
    await client.delete(`/risks/${id}`);
  },

  resolve: async (id: number, note?: string): Promise<Risk> => {
    return riskApi.update(id, { isResolved: true, resolutionNote: note });
  },
};
